src/firebase-db.ts

```diff
@@ -494,3 +497,72 @@ export const getLegacyLinkedInteractiveInfo = (embeddableRefId: string, laraData
       });
     });
 };
+
+export const getApRun = async (sequenceActivity?: string|null) => {
+  if (!portalData) {
+    throw new Error("Must set portal data first");
+  }
+  let query: firebase.firestore.Query = app.firestore().collection(apRunsPath());
+
+  if (portalData.type === "authenticated") {     // logged in user
+    query = query
+      .where("platform_id", "==", portalData.platformId)
+      .where("resource_url", "==", portalData.resourceUrl)
+      .where("context_id", "==", portalData.contextId)
+      .where("platform_user_id", "==", portalData.platformUserId.toString());
+  } else {
+    query = query.where("run_key", "==", portalData.runKey);
+  }
+
+  if (sequenceActivity) {
+    query = query.where("sequence_activity", "==", sequenceActivity);
+  }
+
+  const doc = await query.get();
+  if (doc.empty) {
+    return null;
+  }
+
+  return {id: doc.docs[0].id, data: doc.docs[0].data() as IApRun};
+};
+
+export const createOrUpdateApRun = async ({sequenceActivity, pageId}: {sequenceActivity?: string, pageId: number}) => {
+  if (!portalData) {
+    throw new Error("Must set portal data first");
+  }
+
+  let apRun: IApRun;
+  const existingApRun = await getApRun(sequenceActivity);
+
+  const common: IBaseApRun = {
+    sequence_activity: sequenceActivity || null,
+    page_id: pageId,
+    created: existingApRun?.data.created || utcString(),
+    updated: utcString(),
+  };
+
+  if (portalData.type === "authenticated") {
+    apRun = {
+      type: "authenticated",
+      platform_id: portalData.platformId,
+      platform_user_id: portalData.platformUserId,
+      context_id: portalData.contextId,
+      resource_url: portalData.resourceUrl,
+      ...common
+    };
+  } else {
+    apRun = {
+      type: "anonymous",
+      run_key: portalData.runKey,
+      ...common
+    };
+  }
+
+  const firestoreSetPromise = app.firestore()
+    .doc(apRunsPath(existingApRun?.id))
```

Comment:
Now that I have it on GitHub looking at the diffs this might be buggy - it will overwrite an existing document with a new sequence_activity if that is the first time the sequence_activity is set.